enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,461 @@
+  - "@pacevedom, MicroShift contributor"
+  - "@eslutsky, MicroShift contributor"
+  - "@copejon, MicroShift contributor"
+approvers:
+  - "@jerpeter1, MicroShift Staff Engineer"
+api-approvers:
+  - none
+creation-date: 2025-05-13
+last-updated: 2025-05-13
+tracking-link:
+  - https://issues.redhat.com/browse/OCPSTRAT-2031
+---
+
+# Generic Device Plugin for MicroShift
+
+## Summary
+
+[Generic Device Plugin](https://github.com/squat/generic-device-plugin) is a
+solution for attaching devices not needing special initialization
+(e.g. serial, USB, certain video devices) to containers. The goal of this enhancement is to
+outline its adoption for MicroShift.
+
+## Motivation
+
+MicroShift is built for edge devices, which commonly use peripherals
+like sensors and adapters.
+Currently, MicroShift does not provide any solution to pass these peripherals
+through to Pods without requiring elevated privileges.
+We want to provide a solution for attaching simple devices (not requiring any special
+initialization) to Pods without potentially compromising on security.
+
+### User Stories
+
+* As a MicroShift device administrator, I want to attach simple devices to Pods
+  without elevating privileges (e.g. mounting `/dev`, adding extraneous capabilities).
+
+### Goals
+
+* Include functionality of
+  [Generic Device Plugin](https://github.com/squat/generic-device-plugin)
+  in MicroShift
+
+### Non-Goals
+
+* Adapt Generic Device Plugin (e.g. create Operator) for OpenShift or other Kubernetes distribution
+
+## Proposal
+
+* Import Generic Device Plugin code into MicroShift binary
+* Integrate the Generic Device Plugin's configuration into MicroShift config
+  and add a toggle option
+
+### Workflow Description
+
+**User** is a human user responsible for setting up and managing Edge Devices.
+
+1. The user enables the Generic Device Plugin using MicroShift configuration
+1. Optionally, the user modifies the configuration of Generic Device Plugin, 
+   for example changes device groups and paths
+1. The user (re)starts MicroShift
+1. MicroShift starts the Generic Device Plugin controller which creates plugin
+   goroutines for each device group.
+1. Plugin goroutines detect the devices and report them with the Kubelet
+1. Kubelet updates the Node's status with new devices for scheduler to be aware of Node's capabilities
+1. User deploys a workload utilizing some device
+   (could be at any point, it won't be scheduled/started until the device is available)
+
+In order for workload to get the device attached, the **container** specification
+must include the device in its limits. Below is an example of StatefulSet which
+references a device in a container's limits.
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: WORKLOAD
+  namespace: NAMESPACE
+spec:
+  template:
+    spec:
+      containers:
+      - name: CONTAINER
+        resources:
+          limits:
+            device.microshift.io/serial: "1"
+```
```

Comment:
What happens if the generic device plugin pod restarts? How are the device states and allocations maintained?

Reply:
There will be no pod, GDP will be part of MicroShift binary and if that restarts, it restarts along with kubelet.

But if it were a Pod, I don't think it would impact greatly, because it's actually not the device plugin that manages the mounting of the devices. It's not even a broker or a passthrough.
AFAIK device plugin starts, registers to kubelet, kubelet ask "what do you have", device plugin responds. Then, if some container requests a device, kubelet says "prepare me this device and tell me more", so the device plugin responds with path to a device on the host, so the kubelet knows what information to include when requesting CRIO to create a container.

https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-implementation

Reply:
sorry for the confusion, what i meant was the pod that was allocated with the device, what would happen if it restarts ?